### ISSUE TITLE
[BUG / possible BUG] - Cloudflare TLS 1.0 / TLS 1.1 discoverable but disabled.

### ISSUE THREAD

**Issue by gord0b** [drwetter/testssl.sh#2035]
**Command line / docker command to reproduce**
testssl.sh --openssl=/usr/bin/openssl --ip one -oH app.test.testing.html -oJ app.test.testing.json https://app.test.testing
_PS: removed original testing url._

curl -o /dev/null -L -v -s https://app.test.testing --tls-max 1.0
curl -o /dev/null -L -v -s https://app.test.testing--tls-max 1.1
_**Error returned from above commands:**_ * error:141E70BF:SSL routines:tls_construct_client_hello:no protocols available
**_Reference:_** https://developers.cloudflare.com/ssl/edge-certificates/additional-options/minimum-tls

**Expected behavior**
Example Issue: Even if a minimum TLS version of TLS 1.2 is set in Cloudflare, the following is still reported by TestSSL scan below, when scanning a host on Cloudflare.
 **TLS 1      offered (deprecated)
 TLS 1.1    offered (deprecated)**
 TLS 1.2    offered (OK)
 TLS 1.3    offered (OK): final

Cloudflare allows minimum TLS version to be set, as described in https://support.cloudflare.com/hc/en-us/articles/205043158-Cloudflare-and-PCI-Compliance

**Your system (please complete the following information):**
 - OS: ``awk -F\" '/PRETTY_NAME/ { print $2 }' /etc/os-release``: Ubuntu 20.04.1 LTS
 - Platform: ``uname -srm``: Linux 5.4.0-29-generic x86_64
 - Version: ``testssl.sh -b 2>/dev/null | head -4 | tail -2`` 
testssl.sh       3.1dev from https://testssl.sh/dev/, (79df758 2021-10-20 08:58:46 -- )
 - Version if running from git repo: ``git log | head -1`` 
commit 9683af44a23ce042b08bb966eb4f2c412ec3349a
Merge: d496c0d b6b5a67
Author: Dirk Wetter <dirk@testssl.sh>
Date:   Mon Oct 18 16:52:35 2021 +0200
 - OpenSSL: ``testssl.sh -b 2>/dev/null | awk -F':' '/openssl/ { print $2}'``: 
./bin/openssl.Linux.x86_64
From testssl script: Using "OpenSSL 1.1.1f  31 Mar 2020" [~98 ciphers]  :/usr/bin/openssl

**Additional context**
_None_

**Comment by dcooper16** [comment]
Hello @gord0b,

In this case I believe that testssl.sh is providing the correct results. I tried the `curl` command that you specified and got the same results, but Wireshark seems to show that that a TLS ClientHello is never sent, so that failure is happening on the client not on the server. This seems to be confirmed by https://github.com/curl/curl/issues/5356, which suggests that the failure has to do with the version of OpenSSL shipped with Ubuntu 20.04.

On the other hand, if I try to connect to the server using the OpenSSL that is shipped with testssl.sh (which definitely supports TLSv1), the connection is successful:
```
> printf -- %b 'GET / HTTP/1.1\r\nHost: app.olympusdao.finance\r\nUser-Agent: TLS tester from https://testssl.sh/dev/\r\nAccept-Encoding: identity\r\nAccept: text/*\r\nConnection: Close\r\n\r\n' | ./bin/openssl.Linux.x86_64 s_client -quiet -connect 37.19.206.70:443 -servername app.olympusdao.finance -no_comp -tls1                  
             
depth=2 C = US, O = Internet Security Research Group, CN = ISRG Root X1
verify return:1
depth=1 C = US, O = Let's Encrypt, CN = R3
verify return:1
depth=0 CN = app.olympusdao.finance
verify return:1
HTTP/1.1 200 OK
Date: Mon, 08 Nov 2021 12:45:54 GMT
Content-Type: text/html
Transfer-Encoding: chunked
Connection: close
Vary: Accept-Encoding
Server: BunnyCDN-ASB1-759
CDN-PullZone: 531500
CDN-Uid: 070ccd6e-b4b0-4c90-b45a-e26d7534205d
CDN-RequestCountryCode: US
Access-Control-Allow-Headers: Content-Type, Range, User-Agent, X-Requested-With
Access-Control-Allow-Methods: GET,HEAD,OPTIONS
Access-Control-Allow-Origin: *
Access-Control-Expose-Headers: Content-Range, X-Chunked-Output, X-Stream-Output
Access-Control-Max-Age: 86400
Age: 255085
Cache-Control: max-age=10, stale-while-revalidate=600
Expires: Mon, 08 Nov 2021 16:45:54 GMT
Last-Modified: Fri, 05 Nov 2021 13:54:29 GMT
CF-Ray: 6aaee6b078977004-IAD
Strict-Transport-Security: max-age=15724800; includeSubDomains
CF-Cache-Status: HIT
Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
x-cache-status: MISS
x-ipfs-path: /ipfs/bafybeig6phfrgmqqtak5zqvcduvzrhhzeqqyxza6fn3uspe5szwsywhw7u/
X-Request-ID: a944fa0c3dbf44704294d77a9a780d42
CDN-ProxyVer: 1.0
CDN-RequestPullSuccess: True
CDN-RequestPullCode: 200
CDN-CachedAt: 11/08/2021 12:45:54
CDN-EdgeStorageId: 759
CDN-Status: 200
CDN-RequestId: 57147c78fb0e0b21f7003d12aa630b94
CDN-Cache: EXPIRED

b4d
<!doctype html><html lang="en"><head><script>!function(e,t,a,n,g){e[n]=e[n]||[],e[n].push({"gtm.start":(new Date).getTime(),event:"gtm.js"});var m=t.getElementsByTagName(a)[0],r=t.createElement(a);r.async=!0,r.src="https://www.googletagmanager.com/gtm.js?id=GTM-PD2T9BC",m.parentNode.insertBefore(r,m)}(window,document,"script","dataLayer")</script><script async src="https://www.googletagmanager.com/gtag/js?id=UA-196137638-1"></script><script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","UA-196137638-1")</script><meta charset="utf-8"/><link rel="icon" href="./favicon.ico"/><meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1,minimum-scale=1,user-scalable=no"/><meta name="description" content="Olympus utilizes Treasury Reserves to enable long-term price consistency and scarcity within an infinite supply system"/><link rel="apple-touch-icon" href="./logo192.png"/><link rel="manifest" href="./manifest.json"/><title>OlympusDAO</title><link href="./static/css/main.328c1fc4.chunk.css" rel="stylesheet"></head><body><noscript>You need to enable JavaScript to run this app.</noscript><div id="root"></div><script>!function(e){function t(t){for(var n,a,f=t[0],l=t[1],p=t[2],i=0,s=[];i<f.length;i++)a=f[i],Object.prototype.hasOwnProperty.call(o,a)&&o[a]&&s.push(o[a][0]),o[a]=0;for(n in l)Object.prototype.hasOwnProperty.call(l,n)&&(e[n]=l[n]);for(c&&c(t);s.length;)s.shift()();return u.push.apply(u,p||[]),r()}function r(){for(var e,t=0;t<u.length;t++){for(var r=u[t],n=!0,f=1;f<r.length;f++){var l=r[f];0!==o[l]&&(n=!1)}n&&(u.splice(t--,1),e=a(a.s=r[0]))}return e}var n={},o={1:0},u=[];function a(t){if(n[t])return n[t].exports;var r=n[t]={i:t,l:!1,exports:{}};return e[t].call(r.exports,r,r.exports,a),r.l=!0,r.exports}a.m=e,a.c=n,a.d=function(e,t,r){a.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:r})},a.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},a.t=function(e,t){if(1&t&&(e=a(e)),8&t)return e;if(4&t&&"object"==typeof e&&e&&e.__esModule)return e;var r=Object.create(null);if(a.r(r),Object.defineProperty(r,"default",{enumerable:!0,value:e}),2&t&&"string"!=typeof e)for(var n in e)a.d(r,n,function(t){return e[t]}.bind(null,n));return r},a.n=function(e){var t=e&&e.__esModule?function(){return e.default}:function(){return e};return a.d(t,"a",t),t},a.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},a.p="./";var f=this["webpackJsonp@scaffold-eth/react-app"]=this["webpackJsonp@scaffold-eth/react-app"]||[],l=f.push.bind(f);f.push=t,f=f.slice();for(var p=0;p<f.length;p++)t(f[p]);var c=l;r()}([])</script><script src="./static/js/2.07bac877.chunk.js"></script><script src="./static/js/main.201e4d5b.chunk.js"></script></body></html>
0
```

**Comment by drwetter** [comment]
Have you tried @gord0b whether your openssl allows TLS 1.0 in the first place? Using a standard Debian 10 the curl command line doesn't work for me either. the last lines of ``/etc/ssl/openssl.cnf``:

```
[system_default_sect]
MinProtocol = TLSv1.2
CipherString = DEFAULT@SECLEVEL=2
```


Also: app.olympusdao.finance is pointing to a different CDN = Datacamp with different IPs depending on the location. I queried a resolver in Europe or North America.

**Comment by gord0b** [comment]
> 
> 
> Have you tried @gord0b whether your openssl allows TLS 1.0 in the first place? Using a standard Debian 10 the curl command line doesn't work for me either. the last lines of `/etc/ssl/openssl.cnf`:
> 
> ```
> [system_default_sect]
> MinProtocol = TLSv1.2
> CipherString = DEFAULT@SECLEVEL=2
> ```
> 
> Also: app.olympusdao.finance is pointing to a different CDN = Datacamp with different IPs depending on the location. I queried a resolver in Europe or North America.

Thanks, i did make a change to /etc/ssl/openssl.cnf & changed the MinProtocol to TLSv1.0 and the curl command did work, so client side issue it seems.

**Comment by gord0b** [comment]
> 
> 
> Hello @gord0b,
> 
> In this case I believe that testssl.sh is providing the correct results. I tried the `curl` command that you specified and got the same results, but Wireshark seems to show that that a TLS ClientHello is never sent, so that failure is happening on the client not on the server. This seems to be confirmed by [curl/curl#5356](https://github.com/curl/curl/issues/5356), which suggests that the failure has to do with the version of OpenSSL shipped with Ubuntu 20.04.
> 
> On the other hand, if I try to connect to the server using the OpenSSL that is shipped with testssl.sh (which definitely supports TLSv1), the connection is successful:
> 
> ```
> > printf -- %b 'GET / HTTP/1.1\r\nHost: app.olympusdao.finance\r\nUser-Agent: TLS tester from https://testssl.sh/dev/\r\nAccept-Encoding: identity\r\nAccept: text/*\r\nConnection: Close\r\n\r\n' | ./bin/openssl.Linux.x86_64 s_client -quiet -connect 37.19.206.70:443 -servername app.olympusdao.finance -no_comp -tls1                  
>              
> depth=2 C = US, O = Internet Security Research Group, CN = ISRG Root X1
> verify return:1
> depth=1 C = US, O = Let's Encrypt, CN = R3
> verify return:1
> depth=0 CN = app.olympusdao.finance
> verify return:1
> HTTP/1.1 200 OK
> Date: Mon, 08 Nov 2021 12:45:54 GMT
> Content-Type: text/html
> Transfer-Encoding: chunked
> Connection: close
> Vary: Accept-Encoding
> Server: BunnyCDN-ASB1-759
> CDN-PullZone: 531500
> CDN-Uid: 070ccd6e-b4b0-4c90-b45a-e26d7534205d
> CDN-RequestCountryCode: US
> Access-Control-Allow-Headers: Content-Type, Range, User-Agent, X-Requested-With
> Access-Control-Allow-Methods: GET,HEAD,OPTIONS
> Access-Control-Allow-Origin: *
> Access-Control-Expose-Headers: Content-Range, X-Chunked-Output, X-Stream-Output
> Access-Control-Max-Age: 86400
> Age: 255085
> Cache-Control: max-age=10, stale-while-revalidate=600
> Expires: Mon, 08 Nov 2021 16:45:54 GMT
> Last-Modified: Fri, 05 Nov 2021 13:54:29 GMT
> CF-Ray: 6aaee6b078977004-IAD
> Strict-Transport-Security: max-age=15724800; includeSubDomains
> CF-Cache-Status: HIT
> Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
> x-cache-status: MISS
> x-ipfs-path: /ipfs/bafybeig6phfrgmqqtak5zqvcduvzrhhzeqqyxza6fn3uspe5szwsywhw7u/
> X-Request-ID: a944fa0c3dbf44704294d77a9a780d42
> CDN-ProxyVer: 1.0
> CDN-RequestPullSuccess: True
> CDN-RequestPullCode: 200
> CDN-CachedAt: 11/08/2021 12:45:54
> CDN-EdgeStorageId: 759
> CDN-Status: 200
> CDN-RequestId: 57147c78fb0e0b21f7003d12aa630b94
> CDN-Cache: EXPIRED
> 
> b4d
> <!doctype html><html lang="en"><head><script>!function(e,t,a,n,g){e[n]=e[n]||[],e[n].push({"gtm.start":(new Date).getTime(),event:"gtm.js"});var m=t.getElementsByTagName(a)[0],r=t.createElement(a);r.async=!0,r.src="https://www.googletagmanager.com/gtm.js?id=GTM-PD2T9BC",m.parentNode.insertBefore(r,m)}(window,document,"script","dataLayer")</script><script async src="https://www.googletagmanager.com/gtag/js?id=UA-196137638-1"></script><script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","UA-196137638-1")</script><meta charset="utf-8"/><link rel="icon" href="./favicon.ico"/><meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1,minimum-scale=1,user-scalable=no"/><meta name="description" content="Olympus utilizes Treasury Reserves to enable long-term price consistency and scarcity within an infinite supply system"/><link rel="apple-touch-icon" href="./logo192.png"/><link rel="manifest" href="./manifest.json"/><title>OlympusDAO</title><link href="./static/css/main.328c1fc4.chunk.css" rel="stylesheet"></head><body><noscript>You need to enable JavaScript to run this app.</noscript><div id="root"></div><script>!function(e){function t(t){for(var n,a,f=t[0],l=t[1],p=t[2],i=0,s=[];i<f.length;i++)a=f[i],Object.prototype.hasOwnProperty.call(o,a)&&o[a]&&s.push(o[a][0]),o[a]=0;for(n in l)Object.prototype.hasOwnProperty.call(l,n)&&(e[n]=l[n]);for(c&&c(t);s.length;)s.shift()();return u.push.apply(u,p||[]),r()}function r(){for(var e,t=0;t<u.length;t++){for(var r=u[t],n=!0,f=1;f<r.length;f++){var l=r[f];0!==o[l]&&(n=!1)}n&&(u.splice(t--,1),e=a(a.s=r[0]))}return e}var n={},o={1:0},u=[];function a(t){if(n[t])return n[t].exports;var r=n[t]={i:t,l:!1,exports:{}};return e[t].call(r.exports,r,r.exports,a),r.l=!0,r.exports}a.m=e,a.c=n,a.d=function(e,t,r){a.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:r})},a.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},a.t=function(e,t){if(1&t&&(e=a(e)),8&t)return e;if(4&t&&"object"==typeof e&&e&&e.__esModule)return e;var r=Object.create(null);if(a.r(r),Object.defineProperty(r,"default",{enumerable:!0,value:e}),2&t&&"string"!=typeof e)for(var n in e)a.d(r,n,function(t){return e[t]}.bind(null,n));return r},a.n=function(e){var t=e&&e.__esModule?function(){return e.default}:function(){return e};return a.d(t,"a",t),t},a.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},a.p="./";var f=this["webpackJsonp@scaffold-eth/react-app"]=this["webpackJsonp@scaffold-eth/react-app"]||[],l=f.push.bind(f);f.push=t,f=f.slice();for(var p=0;p<f.length;p++)t(f[p]);var c=l;r()}([])</script><script src="./static/js/2.07bac877.chunk.js"></script><script src="./static/js/main.201e4d5b.chunk.js"></script></body></html>
> 0
> ```

Interesting and thank you, it does seem to be a client side issue when testing with curl. Will close.
